Extend ResourceLoader API for retrieval of (all / only filtered) cached paths in (C++, C#, GDScript)

TruelyMostWanted opened the pull request:
Hey, this is my first PR / contribution to the engine 👋

### Description:
This PR extends the `ResourceLoader` by 4 public methods for retrieval of either all or only filtered "cached" resource paths as arrays or dictionaries in C++, GDScript and C#

**Note(s)** 
A Resource is considered cached when **all of the following conditions are true**:
- It must have a `get_reference_count() > 0` (assigned to exported variables via inspector or assigned to any variable via code)
- It must have a valid `resource_path` set (Examples: `res://icon.svg` or `C:/path/to/my/file.png` )

**Usages:**
```js
extends Node
class_name GdScriptTestNode

@export var resources : Array[Resource]

func _ready() -> void:
	var all_entries_dict = ResourceLoader.get_cached_entries()
	print(JSON.stringify(all_entries_dict, "\t"))

	var all_paths_types = ResourceLoader.get_cached_paths_infos(ResourceLoader.FilterTarget.FILTER_TARGET_TYPE_HINT);
	print(JSON.stringify(all_paths_types, "\t"));

	var all_image_entries = ResourceLoader.get_cached_entries_filtered(
		ResourceLoader.FilterTarget.FILTER_TARGET_TYPE_HINT,
		ResourceLoader.FilterComparator.FILTER_COMP_EQUALS,
		"Image",

	);
	print(JSON.stringify(all_image_entries, "\t"))
	
	var all_res_paths_types = ResourceLoader.get_cached_paths_infos_filtered(
		ResourceLoader.FilterTarget.FILTER_TARGET_TYPE_HINT,
		ResourceLoader.FilterTarget.FILTER_TARGET_PATH,
		ResourceLoader.FILTER_COMP_BEGINS_WITH,
		"res://new"
	);
	print(JSON.stringify(all_res_paths_types, "\t"))
```

```cs
using Godot;
using Godot.Collections;

[GlobalClass]
public partial class TestNode : Node
{
    [Export] public Array<Resource> Resources;

    public override void _Ready()
    {
        var allEntriesDict = ResourceLoader.GetCachedEntries();
        GD.Print(Json.Stringify(allEntriesDict, "\t"));
        
        var allPathsTypesDict  = ResourceLoader.GetCachedPathsInfos(ResourceLoader.FilterTarget.TypeHint);
        GD.Print(Json.Stringify(allPathsTypesDict, "\t"));
        
        var allImgPaths  = ResourceLoader.GetCachedEntriesFiltered(
            ResourceLoader.FilterTarget.TypeHint, 
            ResourceLoader.FilterComparator.Equals, 
            "Image",
            100
        );
        GD.Print(Json.Stringify(allImgPaths, "\t"));
        
        var allResPaths  = ResourceLoader.GetCachedPathsInfosFiltered(
            ResourceLoader.FilterTarget.TypeHint, 
            ResourceLoader.FilterTarget.Path, 
            ResourceLoader.FilterComparator.BeginsWith, 
            "res://new",
            100
        );
        GD.Print(Json.Stringify(allResPaths, "\t"));
    }
}
```

### Result:
With the above GDScript/C# and my list of resources (on the left) assigned to export variables i get the following result:
<img width="1127" height="670" alt="image" src="https://github.com/user-attachments/assets/2e3c130c-2770-4ee5-bcb7-a79914d5be4e" />

### Documentation with Examples In-Engine
<img width="616" height="806" alt="image" src="https://github.com/user-attachments/assets/debd4be8-a163-4b80-8b7c-262408f6942e" />

I hope this helps a lot of your projects out! ✨
Feel free to give me any feedback on the idea, the PR, the description/text of the PR and such

* *Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/13523*

TruelyMostWanted committed:
Extend ResourceLoader Retrieval capabilities

Now you can retrieve all cached paths or only those with certain typeHint, pathPrefix, fileNamePrefix or fileExtension (C++/C#/GDScript)

TruelyMostWanted committed:
Added missing parameter names

TruelyMostWanted committed:
Merge branch 'godotengine:master' into master

TruelyMostWanted committed:
Merge branch 'master' of https://github.com/TruelyMostWanted/yours_truely_godot

TruelyMostWanted committed:
Removed empty lines for clang-format

TruelyMostWanted committed:
Added documentation for ResourceLoader methods

TruelyMostWanted committed:
Merge branch 'godotengine:master' into master

TruelyMostWanted committed:
Finalized "Extended ResourceLoader API" for retrieval queries on cached resources + Updated Documentation with Examples

TruelyMostWanted committed:
Documentation fixed using --doctool

TruelyMostWanted committed:
Reformatted and Cleaned Up Code

TruelyMostWanted committed:
Merge branch 'godotengine:master' into master

TruelyMostWanted committed:
Merge branch 'master' of https://github.com/TruelyMostWanted/yours_truely_godot

TruelyMostWanted committed:
Fixed Documentation, [codeblock] unclosed

AThousandShips commented on core/core_bind.cpp:
```diff
@@ -68,6 +67,18 @@ Ref<Resource> ResourceLoader::load_threaded_get(const String &p_path) {
 	Ref<Resource> res = ::ResourceLoader::load_threaded_get(p_path, &error);
 	return res;
 }
+PackedStringArray ResourceLoader::get_cached_paths() {
```
```suggestion

PackedStringArray ResourceLoader::get_cached_paths() {
```
For all you added, please leave an empty line between methods in the `cpp` files

AThousandShips commented on core/core_bind.h:
```diff
@@ -66,12 +66,37 @@ class ResourceLoader : public Object {
 	};
 
+	enum FilterTarget {
+		FILTER_TARGET_TYPE_HINT = 0,
```
```suggestion
		FILTER_TARGET_TYPE_HINT,
```
Redundant

AThousandShips commented on core/io/resource.h:
```diff
@@ -210,6 +210,46 @@ class ResourceCache {
+		for (const auto &elem : resources) {
+			const String &path = elem.key;
+			Resource *res_ptr = elem.value;
+			Ref<Resource> res = Ref<Resource>(res_ptr);
```
```suggestion
			Ref<Resource> res = res_ptr;
```
Redundant, for both

AThousandShips commented on core/io/resource.h:
```diff
@@ -210,6 +210,46 @@ class ResourceCache {
+			Resource *res_ptr = elem.value;
+			Ref<Resource> res = Ref<Resource>(res_ptr);
+
+			// Prüfen, ob das Predicate für dieses Element true zurückgibt.
```
Please write comments in English

AThousandShips commented on core/io/resource.h:
```diff
@@ -210,6 +210,46 @@ class ResourceCache {
+	}
+
+	template <typename PredicateDict>
+	static Dictionary _get_all_paths_filtered_dict(PredicateDict pred) {
```
```suggestion
	static Dictionary _get_all_paths_filtered_dict(PredicateDict p_pred) {
```

AThousandShips commented on core/io/resource.h:
```diff
@@ -210,6 +210,46 @@ class ResourceCache {
+	template <typename PredicateDict>
+	static Dictionary _get_all_paths_filtered_dict(PredicateDict pred) {
+		MutexLock mutex_lock(lock);
+		Dictionary paths_types_dict = Dictionary();
```
```suggestion
		Dictionary paths_types_dict;
```
Redundant

AThousandShips commented on core/io/resource_loader.cpp:
```diff
@@ -513,6 +515,86 @@ String ResourceLoader::_validate_local_path(const String &p_path) {
+		return p_res->get_class();
+	}
+	return "";
+}
```
```suggestion
}

```
All here as well

AThousandShips commented on core/io/resource_loader.h:
```diff
@@ -126,6 +126,25 @@ class ResourceLoader {
+		FILTER_TARGET_BASE_DIR,
+		FILTER_TARGET_RESOURCE_NAME,
+		FILTER_TARGET_RESOURCE_PATH,
+		FILTER_TARGET_RESOURCE_CLASS
```
```suggestion
		FILTER_TARGET_RESOURCE_CLASS,
```

AThousandShips commented on core/io/resource_loader.h:
```diff
@@ -126,6 +126,25 @@ class ResourceLoader {
+	};
+
+	enum FilterComparator {
+		FILTER_COMP_EQUALS = 0,
```
```suggestion
		FILTER_COMP_EQUALS,
```

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -32,6 +32,98 @@
+				Returns a [Dictionary] of cached resource paths that match a single filter condition in the form of { "res://icon.svg": "Image" }
+				The [param target] selects the attribute and [param comparator] defines comparison against [param value].
+				Allowed [param target]: FILTER_TARGET_TYPE_HINT, FILTER_TARGET_PATH, FILTER_TARGET_FILE_NAME, FILTER_TARGET_FILE_EXTENSION, FILTER_TARGET_BASE_DIR, FILTER_TARGET_RESOURCE_NAME, FILTER_TARGET_RESOURCE_PATH, FILTER_TARGET_RESOURCE_CLASS.
+				Allowed [param comparator]: FILTER_COMP_EQUALS, FILTER_COMP_NOT_EQUAL, FILTER_COMP_BEGINS_WITH,	FILTER_COMP_ENDS_WITH, FILTER_COMP_CONTAINS.
```
```suggestion
				Allowed [param comparator]: FILTER_COMP_EQUALS, FILTER_COMP_NOT_EQUAL, FILTER_COMP_BEGINS_WITH, FILTER_COMP_ENDS_WITH, FILTER_COMP_CONTAINS.
```

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -32,6 +32,98 @@
+				[/codeblock]
+				All resources in the res://assets/ folder:
+				[codeblock]
+				#
```
```suggestion
```

AThousandShips reviewed (commented):
This looks promising and quite minimal in scope, just doing a form pass

AThousandShips reviewed (commented):
This looks promising and quite minimal in scope, just doing a form pass

TruelyMostWanted committed:
Apply Enum suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

TruelyMostWanted committed:
Restored Empty lines

TruelyMostWanted committed:
Added lines between created methods

TruelyMostWanted committed:
Apply suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

AThousandShips commented on core/io/resource_loader.cpp:
```diff
@@ -45,6 +45,8 @@
 #include "core/variant/variant_parser.h"
 #include "servers/rendering/rendering_server.h"
 
+#include <algorithm>
```
What is this used for?

TruelyMostWanted committed:
Refactored ResourceLoader+ResourceCache: templates + auto fully removed

TruelyMostWanted committed:
Apply suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

TruelyMostWanted committed:
Apply enum suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

TruelyMostWanted committed:
Apply documentation suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

AThousandShips commented:
Please apply all the suggestions locally and push once to reduce noise, if you weren't a first time contributor this would have triggered several runs of CI unnecessarily

TruelyMostWanted committed:
Apply doc suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -32,6 +32,93 @@
+			<description>
+				Returns all cached resource paths that match a single filter condition as [PackedStringArray]
+				The [param target] selects which attribute is checked, [param comparator] defines how it is compared to
+				[param value].
```
```suggestion
				Returns all cached resource paths that match a single filter condition as a [PackedStringArray]. The [param target] selects which attribute is checked, [param comparator] defines how it is compared to [param value].
```

TruelyMostWanted committed:
Updated documentation with indents and methods references, removed not needed descriptions

TruelyMostWanted committed:
Merge branch 'resourceloader-get-and-filter-paths' of https://github.com/TruelyMostWanted/yours_truely_godot into resourceloader-get-and-filter-paths

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -179,5 +258,45 @@
+			Target attribute is the full resource path, e.g. [code]"res://assets/ui/button.png"[/code].
+		</constant>
+		<constant name="FILTER_TARGET_FILE_NAME" value="2" enum="FilterTarget">
+			Target attribute is the file name part only (without directories), e.g. "button.png". [method String.get_file]
```
```suggestion
			Target attribute is the file name part only (without directories), e.g. "button.png" ([method String.get_file]).
```
For all

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -179,5 +258,45 @@
+			Target attribute is the resource's user-defined name
+		</constant>
+		<constant name="FILTER_TARGET_RESOURCE_PATH" value="6" enum="FilterTarget">
+			Target attribute is the resource's internal path
```
```suggestion
			Target attribute is the resource's internal path.
```

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -179,5 +258,45 @@
+			Target attribute is the base directory of the resource path, e.g. "res://assets/ui" [method String.get_base_dir]
+		</constant>
+		<constant name="FILTER_TARGET_RESOURCE_NAME" value="5" enum="FilterTarget">
+			Target attribute is the resource's user-defined name
```
```suggestion
			Target attribute is the resource's user-defined name.
```

TruelyMostWanted commented:
> Please apply all the suggestions locally and push once to reduce noise, if you weren't a first time contributor this would have triggered several runs of CI unnecessarily

Oh thats really good to know!
For me this was the first time having an actual code review on GitHub, so once again something learned, thanks!
The next thing i have to learn is rebasing commits into one. Reducing the commit history as seen above

But aside from that, my commits are done @AThousandShips

TruelyMostWanted committed:
Apply suggestions from code review

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

TruelyMostWanted committed:
Apply final suggestions from code review

TruelyMostWanted reviewed (commented):
Code has been updated and all comments resolved.
Even though it shows me a (2) towards "Review changes" i dont see any further ones with status "Pending" or where i have the button to "Add changes to batch"

AThousandShips commented:
Check the "changed files" and you'll see the changes that are left there, in `resource.h` and `core_bind.cpp`

I'll do another review tomorrow

TruelyMostWanted commented:
> Check the "changed files" and you'll see the changes that are left there, in `resource.h` and `core_bind.cpp`
> 
> I'll do another review tomorrow

Alrighty, thanks for your help though! Appreciate it very much!

TruelyMostWanted committed:
Fixed whitespaces

TruelyMostWanted committed:
Polished and improved code/docs

Added mutex lock for thread safety

Fixed Tags Usage in Docs

Adjusted paragraphs

Improved readability by added more references

TruelyMostWanted committed:
Merge branch 'resourceloader-get-and-filter-paths' of https://github.com/TruelyMostWanted/yours_truely_godot into resourceloader-get-and-filter-paths

TruelyMostWanted commented:
Yey, my first rebase of commits worked, nice!
Now i get the hang of it

- Commit all i want / as much as i need to locally but don't push
- Then git rebase all commits into a single Squash Commit with a Reworded Message
- Then when everything's done push the Squashed Commit 

(Alternatively: Open the PR only when the feature is done. But as there can always happen code reviews and discussions, follow the 3 steps above to reduce CI/GitHub action invocations)

Now i'm really done committing on this feature

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -32,6 +32,86 @@
+		<method name="get_cached_paths_typed">
+			<return type="Dictionary" />
+			<description>
+				Returns a [Dictionary] of all cached [Resource] paths as in the form of [code]{ "res://icon.svg": "Image" }[/code]
```
```suggestion
				Returns a [Dictionary] of all cached [Resource] paths using the format [code]{ "res://icon.svg": "Image" }[/code].
```

TruelyMostWanted committed:
Updated exposed ResourceLoader API methods + docs

AThousandShips commented on core/io/resource_loader.h:
```diff
@@ -223,7 +247,16 @@ class ResourceLoader {
 
+	static const Variant _get_cached_info(const FilterTarget &p_target, const String &p_path, const Ref<Resource> &p_res);
+	static bool _is_info_matching_filter(const Variant &p_attribute_value, const FilterComparator &p_filter, const Variant &p_value);
+	static bool _is_key_value_matching_filter(const String &p_path, const Ref<Resource> &p_res, const FilterTarget &p_target, const FilterComparator &p_comparator, const Variant &p_value);
```
```suggestion
	static const Variant _get_cached_info(FilterTarget p_target, const String &p_path, const Ref<Resource> &p_res);
	static bool _is_info_matching_filter(const Variant &p_attribute_value, FilterComparator p_filter, const Variant &p_value);
	static bool _is_key_value_matching_filter(const String &p_path, const Ref<Resource> &p_res, FilterTarget p_target, FilterComparator p_comparator, const Variant &p_value);
```

AThousandShips commented on core/io/resource_loader.h:
```diff
@@ -223,7 +247,16 @@ class ResourceLoader {
+	static Dictionary get_cached_entries();
+	static Dictionary get_cached_paths_infos(const FilterTarget &p_info);
+	static Dictionary get_cached_entries_filtered(const FilterTarget &p_target, const FilterComparator &p_comparator, const Variant &p_value, const int32_t &p_limit);
+	static Dictionary get_cached_paths_infos_filtered(const FilterTarget &p_info, const FilterTarget &p_target, const FilterComparator &p_comparator, const Variant &p_value, const int32_t &p_limit);
```
```suggestion
	static Dictionary get_cached_paths_infos(FilterTarget p_info);
	static Dictionary get_cached_entries_filtered(FilterTarget p_target, FilterComparator p_comparator, const Variant &p_value, int32_t p_limit);
	static Dictionary get_cached_paths_infos_filtered(FilterTarget p_info, FilterTarget p_target, FilterComparator p_comparator, const Variant &p_value, int32_t p_limit);
```

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -32,6 +32,106 @@
+			<param index="3" name="value" type="Variant" />
+			<param index="4" name="limit" type="int" />
+			<description>
+				Returns filtered entries of the ResourceCache as a [Dictionary] with key-value-pairs of [String] and [Variant]
```
```suggestion
				Returns filtered entries of the ResourceCache as a [Dictionary] with key-value-pairs of [String] and [Variant].
```

AThousandShips commented on doc/classes/ResourceLoader.xml:
```diff
@@ -179,5 +279,55 @@
+		<constant name="FILTER_TARGET_RESOURCE_CLASS" value="7" enum="FilterTarget">
+			Target attribute is the resource's class name, e.g., [Texture2D], [Script], [AudioStreamMP3].
+		</constant>
+		<constant name="FILTER_TARGET_REFERENCE_COUNT" value="8" enum="FilterTarget">
```
Needs a description
